contracts/extensions/AmountGetterWithFee.sol

```diff
@@ -0,0 +1,113 @@
+
+/// @title Price getter contract that adds fee calculation
+contract AmountGetterWithFee is AmountGetterBase {
+    /// @dev Allows fees in range [1e-5, 0.65535]
+    uint256 internal constant _FEE_BASE = 1e5;
+    uint256 internal constant _DISCOUNT_BASE = 100;
+
+    /**
+     * @dev Calculates makingAmount with fee.
+     */
+    function _getMakingAmount(
+        IOrderMixin.Order calldata order,
+        bytes calldata extension,
+        bytes32 orderHash,
+        address taker,
+        uint256 takingAmount,
+        uint256 remainingMakingAmount,
+        bytes calldata extraData
+    ) internal view virtual override returns (uint256) {
+        unchecked {
+            (, uint256 integratorFee, uint256 resolverFee, bytes calldata tail) = _parseFeeData(extraData, taker, _isWhitelistedGetterImpl);
+            return Math.mulDiv(
+                super._getMakingAmount(order, extension, orderHash, taker, takingAmount, remainingMakingAmount, tail),
+                _FEE_BASE,
+                _FEE_BASE + integratorFee + resolverFee
+            );
+        }
+    }
+
+    /**
+     * @dev Calculates takingAmount with fee.
+     */
+    function _getTakingAmount(
+        IOrderMixin.Order calldata order,
+        bytes calldata extension,
+        bytes32 orderHash,
+        address taker,
+        uint256 makingAmount,
+        uint256 remainingMakingAmount,
+        bytes calldata extraData
+    ) internal view virtual override returns (uint256) {
+        unchecked {
+            (, uint256 integratorFee, uint256 resolverFee, bytes calldata tail) = _parseFeeData(extraData, taker, _isWhitelistedGetterImpl);
+            return Math.mulDiv(
+                super._getTakingAmount(order, extension, orderHash, taker, makingAmount, remainingMakingAmount, tail),
+                _FEE_BASE + integratorFee + resolverFee,
+                _FEE_BASE,
+                Math.Rounding.Ceil
+            );
+        }
+    }
+
+    /**
+     * @dev `extraData` consists of:
+     * 2 bytes — integrator fee percentage (in 1e5)
+     * 2 bytes — resolver fee percentage (in 1e5)
+     * 1 byte - whitelist discount numerator (in 1e2)
+     * bytes — whitelist structure determined by `_isWhitelisted` implementation
+     * bytes — custom data (optional)
+     * @param _isWhitelisted internal function to parse and check whitelist
+     */
+    function _parseFeeData(
+        bytes calldata extraData,
+        address taker,
+        function (bytes calldata, address) internal view returns (bool, bytes calldata) _isWhitelisted
+    ) internal view returns (bool isWhitelisted, uint256 integratorFee, uint256 resolverFee, bytes calldata tail) {
+        unchecked {
+            integratorFee = uint256(uint16(bytes2(extraData)));
+            resolverFee = uint256(uint16(bytes2(extraData[2:])));
+            uint256 whitelistDiscountNumerator = uint256(uint8(bytes1(extraData[4:])));
+            (isWhitelisted, tail) = _isWhitelisted(extraData[5:], taker);
```

Comment:
Do we need unused `tail` here?

Reply:
Yep, it is used as a named return argument.